armory/art_experimental/attacks/kenansville_dft.py

```diff
@@ -0,0 +1,93 @@
+import numpy as np
+
+
+class KenansvilleDFT:
+    def __init__(
+        self,
+        estimator,
+        sample_rate=16000,
+        snr_db=100,
+        partial_attack=False,
+        attack_len=500,
+        attack_prob=0.5,
+        targeted=False,
+    ):
+        """
+        This DFT attack is a variant of the one described in https://arxiv.org/abs/1910.05262.
+        In the paper, the attack assumed to have knowledge of word or phoneme locations
+        in the input. The attack implemented here assumes complete blackbox knowledge,
+        so the only options are: 1) modified the whole input or 2) modified subsequences
+        of the input with some probability.
+
+        param sample_rate: sample rate in Hz of inputs
+        param estimator: not used but necessary for interoperability with Armory/ART
+        param snr_db: the minimum SNR (in dB) to maintain
+        type snr_db: 'float'
+        param partial_attack: boolean to indicate if subsequences of the input are to be modified
+        param attack_len: length of subsequences to attack. Valid when partial_attack = True
+        param attack_prob: probability each subsequence will be attacked. Valid when partial_attack = True
+        param targeted: not used but necessary for interoperability with Armory
+        """
+        self.sample_rate = sample_rate
+        self.snr_db = snr_db
+        self.targeted = targeted
+        self.partial_attack = partial_attack
+        self.attack_len = attack_len
+        self.attack_prob = attack_prob
+
+    def _attack(self, x):
+        x_fft = np.fft.fft(x)
+        x_psd = np.abs(x_fft) ** 2
+        # sort by frequencies with increasing power
+        x_psd_ind = np.argsort(x_psd)
+        signal_db = 10 * np.log10(np.sum(x_psd))
+
+        """
+        The goal of the following search is to find all the
+        low power frequencies that can be discarded while
+        maintaining a minimum SNR.
+
+        If desired, the following coarse and fine search could
+        be replaced with binary search for faster convergence.
+        """
+        # coarse search
+        id = 2  # real signals have symmetry in positive and negative frequencies
+        noise = np.sum(x_psd[x_psd_ind[:id]])
+        noise_db = 10 * np.log10(noise)
+        while signal_db - noise_db > self.snr_db:
+            id *= 2
+            noise = np.sum(x_psd[x_psd_ind[:id]])
+            noise_db = 10 * np.log10(noise)
+
+        # fine search
+        id = int(id / 2)
+        noise = np.sum(x_psd[x_psd_ind[:id]])
+        noise_db = 10 * np.log10(noise)
+        while signal_db - noise_db > self.snr_db:
```

Comment:
Similarly, if this loop is never entered,  we should not subtract 2 at the next step.  As a simple example, assume the case where neither the coarse or fine searches are entered because `self.snr_db` is too high.  Then, `id` will be 1 before this loop, and `id-2` will become negative, causing most values to be zeroed out.

Reply:
Pushed a new update to address high snr case